File: Python/Strings/L_minimumPalindrome.py

def shortestPalindrome( s: str) -> str:
    count = kmp(s[::-1], s)
    return s[count:][::-1] + s
def kmp(txt: str, patt: str) -> int:
    new_string = patt + '#' + txt
    pi = [0] * len(new_string)
    i = 1
    k = 0
    while i < len(new_string):
        if new_string[i] == new_string[k]:
            k += 1
            pi[i] = k
            i += 1
        else:
            if k > 0:
                k = pi[k - 1]
            else:
                pi[i] = 0
                i += 1
    return pi[-1]

File: JavaScript/String/L_minimumPalindrome.js
var shortestPalindrome = function (s) {
    let count = kmp(s.split("").reverse().join(""), s);
    return s.slice(count).split("").reverse().join("") + s;
};

var kmp = function (txt, patt) {
    let newString = patt + "#" + txt;
    let pi = new Array(newString.length).fill(0);
    let i = 1,
        k = 0;
    while (i < newString.length) {
        if (newString[i] === newString[k]) {
            k++;
            pi[i] = k;
            i++;
        } else {
            if (k > 0) {
                k = pi[k - 1];
            } else {
                pi[i] = 0;
                i++;
            }
        }
    }
    return pi[newString.length - 1];
};
